### PR TITLE
Initial stack implementation

### PR DESCRIPTION
**Pull request by phil9909**
## Summary

Based on the tiny variant: https://github.com/paketo-buildpacks/jammy-tiny-stack/tree/e9a41436b06d3c3fa2960980ea98be36796ab3b6

Changes:
- Renamed
  - search and replace `tiny` with `base`
  - change stack id from `io.buildpacks.stacks.jammy.base` to `io.buildpacks.stacks.jammy` as suggested in #1
- Updated README
- Updated description
- Recreated the `Dockerfile` of the run image based on the build image `Dockerfile`
- Added some packages from bionic base stack
- Updated the integration test (use a simple java app instead of go)

## Use Cases

See #1

I'm not sure if it fixes the issue completely and if I should add a `Closes` keyword to the description.

## Checklist

* [x] I have viewed, signed, and submitted the Contributor License Agreement.
* [ ] I have linked issue(s) that this PR should close using keywords or the Github UI (See [docs](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue))
* [x] I have added an integration test, if necessary.
* [x] I have reviewed the [styleguide](https://github.com/paketo-buildpacks/community/blob/main/STYLEGUIDE.md) for guidance on my code quality.
* [ ] I'm happy with the commit history on this PR (I have rebased/squashed as needed).

### CODE REVIEW

**Comment by robdimsdale** [comment]
@phil9909 awesome! So quick! Thanks for submitting this.

Before I dig into a review, I've noticed that the find-and-replace didn't work inside the `.github` directory - the workflows point to actions on the `jammy-tiny-stack` repository. Could you replace `tiny` with `base` inside the `.github` directory too?

File: stack/run.Dockerfile
Comment (robdimsdale) [inline stack/run.Dockerfile:23]: This `echo 'PRETTY_NAME='` results in a duplicate key. Can we instead use the `sed` that we do above?

File: stack/stack.toml
Comment (robdimsdale) [inline stack/stack.toml:53]: I don't think we need base-files since we are basing the image off of `ubuntu:jammy`. In the tiny stack we need this because we are basing the image off of `SCRATCH`

File: metadata_test.go
Comment (robdimsdale) [inline metadata_test.go:73]: Shouldn't this label have the value `io.buildpacks.stacks.jammy.base` ?

File: metadata_test.go
Comment (robdimsdale) [inline metadata_test.go:159]: Again, shouldn't this key have the value: `io.buildpacks.stacks.jammy.base`?

File: metadata_test.go
Comment (ryanmoran) [reply metadata_test.go:73]: The [RFC outlines](https://github.com/paketo-buildpacks/rfcs/blob/main/text/stacks/0004-jammy-jellyfish.md#stack-ids) that the stack ID for base and full will be the same, `io.buildpacks.stacks.jammy`. This aligns with what is currently done in Bionic.

File: stack/build.Dockerfile
Comment (robdimsdale) [inline stack/build.Dockerfile:20]: This URL is wrong - it should be `yj-linux-amd64`. We should also ensure that `curl` fails on error. Let's use the following command:

```
RUN curl -sSfL -o /usr/local/bin/yj https://github.com/sclevine/yj/releases/latest/download/yj-linux-amd64 \
  && chmod +x /usr/local/bin/yj
```

Comment (robdimsdale) [reply stack/build.Dockerfile:20]: (from :https://github.com/paketo-buildpacks/stacks/blob/c00defb718858eb3724c71addae362c9c047f068/bionic/dockerfile/build/Dockerfile#L21)

File: metadata_test.go
Comment (phil9909) [reply metadata_test.go:73]: According to the RFC `io.buildpacks.stacks.jammy` is correct  https://github.com/paketo-buildpacks/rfcs/blob/ba9273ded6c65274276677b1ad5db5e61029060d/text/stacks/0004-jammy-jellyfish.md#stack-ids

This is also the case for bionic stacks:
- full id: `io.buildpacks.stacks.bionic`
- base id: `io.buildpacks.stacks.bionic`
- tiny id: `io.paketo.stacks.tiny`

Comment (robdimsdale) [reply metadata_test.go:73]: yes, sorry, I was looking at `tiny` and forgetting that we treat it differently than `base` and `full`.

File: stack/build.Dockerfile
Comment (phil9909) [reply stack/build.Dockerfile:20]: This was copied from https://github.com/paketo-buildpacks/jammy-tiny-stack/blob/e9a41436b06d3c3fa2960980ea98be36796ab3b6/stack/build.Dockerfile#L20-L21

I will create a PR to fix it also there

Comment (robdimsdale) [reply stack/build.Dockerfile:20]: Perfect, thank you. Yes, we copied that command before we fixed it in `bionic` and clearly forgot to port the fix over to `jammy`. Thanks for submitting a PR there, too.

Comment (phil9909) [reply stack/build.Dockerfile:20]: https://github.com/paketo-buildpacks/jammy-tiny-stack/pull/10

File: stack/stack.toml
Comment (ryanmoran) [inline stack/stack.toml:55]: I think this is a addition to this package list that is different than what we have in bionic. Can you explain why we need to include this package here?

File: stack/build.Dockerfile
Comment (ryanmoran) [inline stack/build.Dockerfile:12]: These lines reference `$package_args`, but that doesn't appear to be defined anywhere. The bionic stack image has the following:

https://github.com/paketo-buildpacks/stacks/blob/c00defb718858eb3724c71addae362c9c047f068/bionic/dockerfile/build/Dockerfile#L5

File: stack/run.Dockerfile
Comment (ryanmoran) [inline stack/run.Dockerfile:16]: In bionic, we remove a few other things:
https://github.com/paketo-buildpacks/stacks/blob/c00defb718858eb3724c71addae362c9c047f068/bionic/dockerfile/run/Dockerfile#L17-L21

Should we also be removing those here?

File: stack/stack.toml
Comment (sophiewigmore) [reply stack/stack.toml:55]: Additionally, we prefer the packages in the run image be a subset of the packages for in the build image, so I'd also like to understand if there's rationale for having it in the run image only.

File: stack/build.Dockerfile
Comment (robdimsdale) [reply stack/build.Dockerfile:12]: (fixed in the tiny stack with: https://github.com/paketo-buildpacks/jammy-tiny-stack/pull/13/files)

Comment (robdimsdale) [reply stack/build.Dockerfile:12]: I think we can remove the references, rather than populate the arguments. Digging into each argument in more detail:

* `--allow-downgrades` - this allows apt to downgrade the install version of a package without prompting. This seems not great.
* `--allow-remove-essential` - this allows apt to remove core components of the base/essential packages without prompting. This also seems not great.
* `--allow-change-held-packages` - this allows apt to proceed with changing held packages without prompting. This seems moderately bad, but not awful, as I don't think we hold packages.
* `--no-install-recommends` - this instructs apt to not consider recommended packages as a dependency for installing. I'm not sure I have enough context on this to have strong opinions either way.

See: https://manpages.ubuntu.com/manpages/xenial/en/man8/apt-get.8.html

When [this change was first introduced](https://github.com/paketo-buildpacks/stacks/commit/edde18c36e988ad7da926783e49fdf5ab5a8b94b) these args were added without any particular context or extra information, so it's hard to know what the original thinking was.

Comment (sophiewigmore) [reply stack/build.Dockerfile:12]: @robdimsdale thanks for adding that context. In reference to `--no-install-recommends`, I'm not sure how a package becomes recommended, but I would imagine we wouldn't want to install any packages except the ones we've defined, so I'd say we probably want this flag

Comment (robdimsdale) [reply stack/build.Dockerfile:12]: @sophiewigmore yes, that makes sense. However it's not obvious to me that we want that feature, though, either. I think we rely on transitive dependencies - I don't think we list out the exhaustive, flattened list of libraries. I think we'd have to investigate this further.

That's why I don't have enough context to suggest including or excluding it.

Comment (robdimsdale) [reply stack/build.Dockerfile:12]: https://ubuntu.com/blog/we-reduced-our-docker-images-by-60-with-no-install-recommends suggests that `--no-install-recommends` is probably a good idea.

File: stack/stack.toml
Comment (phil9909) [reply stack/stack.toml:8]: I just took it from https://github.com/paketo-buildpacks/jammy-tiny-stack/blob/4d348c6c13b1b6c815c637a5f824f05ec367c06a/stack/stack.toml#L8

Maybe @fg-j can explain

File: stack/stack.toml
Comment (phil9909) [reply stack/stack.toml:55]: We took that from https://github.com/paketo-buildpacks/jammy-tiny-stack/blob/4d348c6c13b1b6c815c637a5f824f05ec367c06a/stack/stack.toml#L48

Our reasoning was that everything that is part of the `tiny` stack should also be part of the `base` stack.

Comment (robdimsdale) [reply stack/stack.toml:54]: `libc6` is already present on ubuntu. We explicitly add it to `tiny`'s `run` image because that is distroless.

I think it's clearer to remove it from the list.

Verifying that it is already present on `ubuntu:jammy`:
```
 ❯ docker run -it ubuntu:jammy /bin/bash
root@6dea59df0ae8:/# echo "deb http://archive.ubuntu.com/ubuntu jammy main" > /etc/apt/sources.list
root@6dea59df0ae8:/# echo "deb http://archive.ubuntu.com/ubuntu jammy-updates main" >> /etc/apt/sources.list
root@6dea59df0ae8:/# echo "deb http://archive.ubuntu.com/ubuntu jammy-security main" >> /etc/apt/sources.list
root@6dea59df0ae8:/# apt update && apt install libc6
Get:1 http://archive.ubuntu.com/ubuntu jammy InRelease [270 kB]
Get:2 http://archive.ubuntu.com/ubuntu jammy-updates InRelease [109 kB]
Get:3 http://archive.ubuntu.com/ubuntu jammy-security InRelease [110 kB]
Get:4 http://archive.ubuntu.com/ubuntu jammy/main amd64 Packages [1792 kB]
Get:5 http://archive.ubuntu.com/ubuntu jammy-updates/main amd64 Packages [406 kB]
Get:6 http://archive.ubuntu.com/ubuntu jammy-security/main amd64 Packages [238 kB]
Fetched 2926 kB in 2s (1842 kB/s)
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
8 packages can be upgraded. Run 'apt list --upgradable' to see them.
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
libc6 is already the newest version (2.35-0ubuntu3).
0 upgraded, 0 newly installed, 0 to remove and 8 not upgraded.
```

Comment (sophiewigmore) [reply stack/stack.toml:54]: Looking at the [Bionic Base stack](https://github.com/paketo-buildpacks/base-stack-release/blob/main/build-receipt), it looks like `libc6` is on the build and run images, even though it's not explicitly required in the Bionic Base stack package list. 

@phil9909 I think your rationale to keep `tiny` as a subset of `base` makes sense and is a guarantee we should try to uphold. I would like to understand if we should include this package in the Base build image as well for consistency's sake. thoughts @robdimsdale @ryanmoran

Comment (robdimsdale) [reply stack/stack.toml:54]: I think we should evaluate the "package list subset claim" on the resultant images, not the input in `stack.toml`.

I say this because the OS brings in a bunch of packages that we don't explicitly list (and probably shouldn't).

That being said, for everything except `tiny`/`run`, we use the same base image, which therefore will have the same set of packages, so if we treat `tiny`/`run` as a special case, then we can consider a subset of the package list in `stack.toml` as equivalent to a subset of the resultant package list, as long as we are consistent in not re-including packages that ship with the OS, such as `libc6`.

Comment (robdimsdale) [reply stack/stack.toml:54]: Here is a list of all the packages that ship by default with ubuntu jammy:

```
 ❯ docker run -it ubuntu:jammy /bin/bash
root@48163cbb2bb8:/#  dpkg -l | cut -d' ' -f3
adduser
apt
base-files
base-passwd
bash
bsdutils
coreutils
dash
debconf
debianutils
diffutils
dpkg
e2fsprogs
findutils
gcc-12-base:amd64
gpgv
grep
gzip
hostname
init-system-helpers
libacl1:amd64
libapt-pkg6.0:amd64
libattr1:amd64
libaudit-common
libaudit1:amd64
libblkid1:amd64
libbz2-1.0:amd64
libc-bin
libc6:amd64
libcap-ng0:amd64
libcap2:amd64
libcom-err2:amd64
libcrypt1:amd64
libdb5.3:amd64
libdebconfclient0:amd64
libext2fs2:amd64
libffi8:amd64
libgcc-s1:amd64
libgcrypt20:amd64
libgmp10:amd64
libgnutls30:amd64
libgpg-error0:amd64
libgssapi-krb5-2:amd64
libhogweed6:amd64
libidn2-0:amd64
libk5crypto3:amd64
libkeyutils1:amd64
libkrb5-3:amd64
libkrb5support0:amd64
liblz4-1:amd64
liblzma5:amd64
libmount1:amd64
libncurses6:amd64
libncursesw6:amd64
libnettle8:amd64
libnsl2:amd64
libp11-kit0:amd64
libpam-modules:amd64
libpam-modules-bin
libpam-runtime
libpam0g:amd64
libpcre2-8-0:amd64
libpcre3:amd64
libprocps8:amd64
libseccomp2:amd64
libselinux1:amd64
libsemanage-common
libsemanage2:amd64
libsepol2:amd64
libsmartcols1:amd64
libss2:amd64
libssl3:amd64
libstdc++6:amd64
libsystemd0:amd64
libtasn1-6:amd64
libtinfo6:amd64
libtirpc-common
libtirpc3:amd64
libudev1:amd64
libunistring2:amd64
libuuid1:amd64
libxxhash0:amd64
libzstd1:amd64
login
logsave
lsb-base
mawk
mount
ncurses-base
ncurses-bin
passwd
perl-base
procps
sed
sensible-utils
sysvinit-utils
tar
ubuntu-keyring
usrmerge
util-linux
zlib1g:amd64
```

Comment (sophiewigmore) [reply stack/stack.toml:54]: @robdimsdale I hadn't seen your comment when I posted mine-- what you wrote makes sense.

> I think we should evaluate the "package list subset claim" on the resultant images, not the input in stack.toml.

^ +1 on this. 

So the outcome here for @phil9909 is that we should remove `libc6` from the `stack.toml` here, since it will be implicitly included with ubuntu. It's only added to Tiny run image explicitly because the tiny run image is based on the `scratch` image, not `ubuntu:jammy`

**Review by robdimsdale — APPROVED** [review]
Acceptance tests pass locally.

I'm happy to squash when merging.

**Comment by robdimsdale** [comment]
@paketo-buildpacks/stacks-maintainers @ryanmoran I'll hold off merging until you all have a chance to approve.

**Review by sophiewigmore — APPROVED** [review]

**Review by ryanmoran — APPROVED** [review]

**Pull request merged**
